parlai/agents/transformer/polyencoder.py

```diff
@@ -368,19 +505,25 @@ def encode(self, ctxt_tokens, cand_tokens):
             assert len(ctxt_tokens.shape) == 2
             bsz = ctxt_tokens.size(0)
             # get context_representation. Now that depends on the cases.
-            ctxt_out, ctxt_mask, ctxt_pos = self.encoder_ctxt(ctxt_tokens)
-            att_keys = ctxt_out if self.attention_keys == 'context' else ctxt_pos
+            if self.use_image_features:
+                if not isinstance(ctxt_image, torch.Tensor) or ctxt_image.size() != (
+                    bsz,
+                    self.image_features_dim,
+                ):
+                    raise ValueError('Image feature tensor malformed!')
+                ctxt_out, ctxt_mask = self.encoder_ctxt(ctxt_tokens, ctxt_image)
```

Comment:
This is starting to feel like we need `_model_input` or similar, like we have in TorchGeneratorAgent.

Reply:
Yes - I've re-written a few functions to reflect the fact that sometimes we'll be passing in ctxt image features and sometimes we won't